s3prl/downstream/a2a-vc-vctk/vc_evaluate.py

```diff
@@ -0,0 +1,369 @@
+    diff2sum = np.sum((cvt_mcep_dtw_pow - gt_mcep_dtw_pow) ** 2, 1)
+    mcd = np.mean(10.0 / np.log(10.0) * np.sqrt(2 * diff2sum), 0)
+
+    # VAD & DTW based on f0
+    gt_nonsil_f0_idx = np.where(gt_feats["f0"] > 0)[0]
+    gt_mcep_nonsil_f0 = gt_feats["mcep"][gt_nonsil_f0_idx]
+    cvt_nonsil_f0_idx = np.where(cvt_feats["f0"] > 0)[0]
+    cvt_mcep_nonsil_f0 = cvt_feats["mcep"][cvt_nonsil_f0_idx]
+    _, path = fastdtw(cvt_mcep_nonsil_f0, gt_mcep_nonsil_f0, dist=scipy.spatial.distance.euclidean)
+    twf_f0 = np.array(path).T
+
+    # f0RMSE, f0CORR using f0-based DTW
+    cvt_f0_dtw = cvt_feats["f0"][cvt_nonsil_f0_idx][twf_f0[0]]
+    gt_f0_dtw = gt_feats["f0"][gt_nonsil_f0_idx][twf_f0[1]]
+    f0rmse = np.sqrt(np.mean((cvt_f0_dtw - gt_f0_dtw) ** 2))
+    f0corr = scipy.stats.pearsonr(cvt_f0_dtw, gt_f0_dtw)[0]
+
+    # DDUR
+    # energy-based VAD with librosa
+    x_trim, _ = librosa.effects.trim(y=x)
+    y_trim, _ = librosa.effects.trim(y=y)
+    ddur = float(abs(len(x_trim) - len(y_trim)) / fs)
+
+    return mcd, f0rmse, f0corr, ddur
+
+
+################################################################################
+
+# The follow section is related to the calculation of ASR-based metrics
+# Reference: https://github.com/tzuhsien/Voice-conversion-evaluation/blob/master/metrics/character_error_rate/inference.py 
+
+import editdistance as ed
+import jiwer
+import torch
+from transformers import Wav2Vec2ForCTC, Wav2Vec2Tokenizer, Wav2Vec2Processor
+    
+ASR_PRETRAINED_MODEL = "facebook/wav2vec2-large-960h-lv60-self"
+
+def load_asr_model(device):
+    """Load model"""
+    print(f"[INFO]: Load the pre-trained ASR by {ASR_PRETRAINED_MODEL}.")
+    model = Wav2Vec2ForCTC.from_pretrained(ASR_PRETRAINED_MODEL).to(device)
+    tokenizer = Wav2Vec2Tokenizer.from_pretrained(ASR_PRETRAINED_MODEL)
+    models = {"model": model, "tokenizer": tokenizer}
+    return models
+
+
+def normalize_sentence(sentence):
+    """Normalize sentence"""
+    # Convert all characters to upper.
+    sentence = sentence.upper()
+    # Delete punctuations.
+    sentence = jiwer.RemovePunctuation()(sentence)
+    # Remove \n, \t, \r, \x0c.
+    sentence = jiwer.RemoveWhiteSpace(replace_by_space=True)(sentence)
+    # Remove multiple spaces.
+    sentence = jiwer.RemoveMultipleSpaces()(sentence)
+    # Remove white space in two end of string.
+    sentence = jiwer.Strip()(sentence)
+
+    # Convert all characters to upper.
+    sentence = sentence.upper()
+
+    return sentence
+
+
+def calculate_measures(groundtruth, transcription):
+    """Calculate character/word measures (hits, subs, inserts, deletes) for one given sentence"""
+    groundtruth = normalize_sentence(groundtruth)
+    transcription = normalize_sentence(transcription)
+
+    #cer = ed.eval(transcription, groundtruth) / len(groundtruth)
+    c_result = jiwer.compute_measures([c for c in groundtruth if c != " "], [c for c in transcription if c != " "])
```

Comment:
This is problematic with the latest jiwer. Will be nice to use another solution for CER. The same issue allso happens in the a2o recipe.